cloudigrade/util/aws.py

```diff
@@ -220,6 +229,120 @@ def get_ec2_instance(session, instance_id):
     return session.resource('ec2').Instance(instance_id)
 
 
+def get_ami(session, image_id, region):
+    """
+    Return an Amazon Machine Image running on an EC2 instance.
+
+    Args:
+        session (boto3.Session): A temporary session tied to a customer account
+        image_id (str): An AMI ID
+
+    Returns:
+        Image: A boto3 EC2 Image object.
+
+    """
+    return session.resource('ec2', region_name=region).Image(image_id)
+
+
+def get_ami_snapshot_id(ami):
+    """
+    Return the snapshot id from an Image object.
+
+    Args:
+        ami (boto3.resources.factory.ec2.Image): A machine image object
+
+    Returns:
+        string: The snapshot id for the machine image's root volume
+
+    """
+    for mapping in ami.block_device_mappings:
+        # For now we are focusing exclusively on the root device
+        if mapping['DeviceName'] != ami.root_device_name:
+            continue
+        return mapping.get('Ebs', {}).get('SnapshotId', '')
+
+
+def change_snapshot_ownership(session, snapshot_id, region, operation='add'):
+    """
+    Add or remove permissions to a snapshot.
+
+    Args:
+        session (boto3.Session): A temporary session tied to a customer account
+        snapshot_id (str): The id of the snapshot to modify
+        region (str): The region the machine image and snapshot reside in
+        operation (str): Should be one of 'add'|'remove'
+
+    Returns:
+        bool: Whether ownership has been modified
+
+    """
+    acceptable_operations = ('add', 'remove')
+    if operation not in acceptable_operations:
+        raise AwsValueError(_('{o} not one of {ao}').format(
+            o=operation,
+            ao=acceptable_operations)
+        )
+
+    snapshot = session.resource('ec2',
+                                region_name=region).Snapshot(snapshot_id)
+
+    if snapshot.encrypted:
+        logger.warning(_('Snapshot is encrypted. Not copying.'))
+        raise AwsSnapshotEncryptedError
+
+    attribute = 'createVolumePermission'
+    user_id = _get_primary_account_id()
+
+    permission = {
+        operation.capitalize(): [
+            {'UserId': user_id},
+        ]
+    }
+
+    snapshot.modify_attribute(
+        Attribute=attribute,
+        CreateVolumePermission=permission,
+        OperationType=operation,
+        UserIds=[user_id]
+    )
+
+    # The modify call returns None. This is a check to make sure
+    # permissions are added successfully.
+    response = snapshot.describe_attribute(Attribute='createVolumePermission')
+
+    for user in response['CreateVolumePermissions']:
+        if user['UserId'] == user_id:
```

Comment:
I don't think this will work in the "remove" case. Removing a permission, by definition, means there will be no match and, therefore, this will fall through to raising an exception. Given that "remove" is a valid option to pass in, and your docstring doesn't mention raising an exception as valid/expected behavior, I think this is a bug.

IMO, I think you'd be better off refactoring this into two methods: add_owner_to_snapshot() and remove_owner_from_snapshot().